# Breast Cancer prediction using Turi Create

The data for this jupyter notebook was initially obtained from https://archive.ics.uci.edu/ml/datasets/Breast+Cancer+Wisconsin+(Original) 

Data description can be found at https://archive.ics.uci.edu/ml/machine-learning-databases/breast-cancer-wisconsin/breast-cancer-wisconsin.names

Objective 

Predict breast cancer based on a set of cytology features (clump thickness, uniformity of cell size, uniformity of cell shape, marginal adhesion, single epithelial cell size, bare nuclei, bland chromatin, normal nuceloli, mitosis.

Data Preprocessing 

Rows with missing values -"NaN" were deleted. Labels were added in the first row based on the data description. Values in the field 'classs' was replaced as follows 2=0(benign) 4=1 (malignant /cancer). No other pre-processing was done. The data file "BreastCancer2.csv" is present in this same GitHub folder as this Jupyter Notebook.

In [1]:
# importing turi create as tc
import turicreate as tc

In [2]:
#Loading the data from the CSV file and assigning it to the variable data
data =  tc.SFrame('BreastCancer2.csv')

Finished parsing file /Users/johnsonthomas/BreastCancer2.csv

Parsing completed. Parsed 100 lines in 0.078517 secs.

------------------------------------------------------
Inferred types from first 100 line(s) of file as 
column_type_hints=[int,int,int,int,int,int,int,int,int,int,int]
If parsing fails due to incorrect types, you can correct
the inferred type list above and pass it to read_csv in
the column_type_hints argument
------------------------------------------------------


Finished parsing file /Users/johnsonthomas/BreastCancer2.csv

Parsing completed. Parsed 683 lines in 0.011224 secs.

In [3]:
# Splitting the data into training (80% of data) and testing (20% of data) sets.
train_data, test_data = data.random_split(0.8)

In [4]:
# Creating the model. 'target' is the variable we are trying to predict. 'features' are the fields that we 
# use to predict the target
# tc.classifier.create command evalauates the data using different models (like logistic regression, radomforest) 
# and finds the best model. 
# The bst model is finally assigned to the variable 'model'
# In this example RandomForestClassifier was selected as the best model based on validation set performance

model = tc.classifier.create(train_data, target='class',
                             features = ['thickness', 'size','shape','adhesion','single','nuclei',
                                         'chromatin','nucleoli','mitosis'
                                         ])

PROGRESS: Creating a validation set from 5 percent of training data. This may take a while.
          You can set ``validation_set=None`` to disable validation tracking.

PROGRESS: The following methods are available for this type of problem.
PROGRESS: BoostedTreesClassifier, RandomForestClassifier, DecisionTreeClassifier, SVMClassifier, LogisticClassifier
PROGRESS: The returned model will be chosen according to validation accuracy.


Boosted trees classifier:

--------------------------------------------------------

Number of examples          : 505

Number of classes           : 2

Number of feature columns   : 9

Number of unpacked features : 9

+-----------+--------------+-------------------+---------------------+-------------------+---------------------+

| Iteration | Elapsed Time | Training-accuracy | Validation-accuracy | Training-log_loss | Validation-log_loss |

+-----------+--------------+-------------------+---------------------+-------------------+---------------------+

| 1         | 0.030457     | 0.990099          | 0.925926            | 0.460310          | 0.476524            |

| 2         | 0.034567     | 0.992079          | 0.962963            | 0.327368          | 0.344539            |

| 3         | 0.037896     | 0.992079          | 0.962963            | 0.242459          | 0.266023            |

| 4         | 0.040458     | 0.994059          | 0.962963            | 0.183961          | 0.209255            |

| 5         | 0.042430     | 0.994059          | 0.962963            | 0.140719          | 0.180387            |

| 6         | 0.044574     | 0.996040          | 0.962963            | 0.111089          | 0.148529            |

+-----------+--------------+-------------------+---------------------+-------------------+---------------------+

Random forest classifier:

--------------------------------------------------------

Number of examples          : 505

Number of classes           : 2

Number of feature columns   : 9

Number of unpacked features : 9

+-----------+--------------+-------------------+---------------------+-------------------+---------------------+

| Iteration | Elapsed Time | Training-accuracy | Validation-accuracy | Training-log_loss | Validation-log_loss |

+-----------+--------------+-------------------+---------------------+-------------------+---------------------+

| 1         | 0.003705     | 0.986139          | 0.925926            | 0.174063          | 0.248492            |

| 2         | 0.005938     | 0.974257          | 0.962963            | 0.177711          | 0.195962            |

| 3         | 0.008466     | 0.976238          | 1.000000            | 0.176312          | 0.190857            |

| 4         | 0.010812     | 0.976238          | 1.000000            | 0.175467          | 0.182376            |

| 5         | 0.012775     | 0.978218          | 1.000000            | 0.175570          | 0.183180            |

| 6         | 0.015548     | 0.978218          | 1.000000            | 0.173633          | 0.180085            |

+-----------+--------------+-------------------+---------------------+-------------------+---------------------+

Decision tree classifier:

--------------------------------------------------------

Number of examples          : 505

Number of classes           : 2

Number of feature columns   : 9

Number of unpacked features : 9

+-----------+--------------+-------------------+---------------------+-------------------+---------------------+

| Iteration | Elapsed Time | Training-accuracy | Validation-accuracy | Training-log_loss | Validation-log_loss |

+-----------+--------------+-------------------+---------------------+-------------------+---------------------+

| 1         | 0.002788     | 0.990099          | 0.925926            | 0.168637          | 0.208652            |

+-----------+--------------+-------------------+---------------------+-------------------+---------------------+

SVM:

--------------------------------------------------------

Number of examples          : 505

Number of classes           : 2

Number of feature columns   : 9

Number of unpacked features : 9

Number of coefficients    : 10

Starting L-BFGS

--------------------------------------------------------

+-----------+----------+-----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-accuracy | Validation-accuracy |

+-----------+----------+-----------+--------------+-------------------+---------------------+

| 1         | 3        | 0.001980  | 0.010070     | 0.334653          | 0.407407            |

| 2         | 7        | 3.000000  | 0.012959     | 0.930693          | 0.888889            |

| 3         | 8        | 3.000000  | 0.014706     | 0.897030          | 0.851852            |

| 4         | 9        | 3.000000  | 0.015608     | 0.762376          | 0.814815            |

| 5         | 11       | 1.000000  | 0.017522     | 0.908911          | 0.925926            |

| 6         | 12       | 1.000000  | 0.020780     | 0.954455          | 0.962963            |

+-----------+----------+-----------+--------------+-------------------+---------------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Logistic regression:

--------------------------------------------------------

Number of examples          : 505

Number of classes           : 2

Number of feature columns   : 9

Number of unpacked features : 9

Number of coefficients    : 10

Starting Newton Method

--------------------------------------------------------

+-----------+----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Elapsed Time | Training-accuracy | Validation-accuracy |

+-----------+----------+--------------+-------------------+---------------------+

| 1         | 2        | 0.060184     | 0.962376          | 0.962963            |

| 2         | 3        | 0.061182     | 0.966337          | 0.962963            |

| 3         | 4        | 0.061983     | 0.970297          | 0.962963            |

| 4         | 5        | 0.063025     | 0.970297          | 0.962963            |

| 5         | 6        | 0.063747     | 0.974257          | 0.962963            |

| 6         | 7        | 0.064724     | 0.974257          | 0.962963            |

+-----------+----------+--------------+-------------------+---------------------+

SUCCESS: Optimal solution found.

PROGRESS: Model selection based on validation accuracy:
PROGRESS: ---------------------------------------------
PROGRESS: BoostedTreesClassifier          : 0.962962985039
PROGRESS: RandomForestClassifier          : 1.0
PROGRESS: DecisionTreeClassifier          : 0.925925910473
PROGRESS: SVMClassifier                   : 0.962963
PROGRESS: LogisticClassifier              : 0.962963
PROGRESS: ---------------------------------------------
PROGRESS: Selecting RandomForestClassifier based on validation set performance.


In [5]:
# Get predictions using the test data 
predictions = model.classify(test_data)

In [6]:
predictions

Columns:
	class	int
	probability	float

Rows: 151

Data:
+-------+----------------+
| class |  probability   |
+-------+----------------+
|   0   | 0.876951314509 |
|   0   | 0.876951314509 |
|   1   | 0.773324787617 |
|   1   | 0.533553183079 |
|   0   | 0.876951314509 |
|   0   | 0.876951314509 |
|   1   | 0.604591429234 |
|   0   | 0.626685112715 |
|   1   | 0.577280879021 |
|   1   | 0.751432836056 |
+-------+----------------+
[151 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [7]:
# obtain statistical results for the model by model.evaluate method 
results = model.evaluate(test_data)

In [8]:
results

{'accuracy': 0.9536423841059603,
 'auc': 0.986919675755343,
 'confusion_matrix': Columns:
 	target_label	int
 	predicted_label	int
 	count	int
 
 Rows: 4
 
 Data:
 +--------------+-----------------+-------+
 | target_label | predicted_label | count |
 +--------------+-----------------+-------+
 |      1       |        0        |   5   |
 |      1       |        1        |   54  |
 |      0       |        0        |   90  |
 |      0       |        1        |   2   |
 +--------------+-----------------+-------+
 [4 rows x 3 columns],
 'f1_score': 0.9391304347826087,
 'log_loss': 0.22541536105489204,
 'precision': 0.9642857142857143,
 'recall': 0.9152542372881356,
 'roc_curve': Columns:
 	threshold	float
 	fpr	float
 	tpr	float
 	p	int
 	n	int
 
 Rows: 100001
 
 Data:
 +-----------+-----+-----+----+----+
 | threshold | fpr | tpr | p  | n  |
 +-----------+-----+-----+----+----+
 |    0.0    | 1.0 | 1.0 | 59 | 92 |
 |   1e-05   | 1.0 | 1.0 | 59 | 92 |
 |   2e-05   | 1.0 | 1.0 | 59 | 92 |
 |

Model has an accuracy of  95.36%
AUC 98.69%


# End of Jupyter notebook